valora-inc/wallet
# fix(statsig): Refactor Statsig initialization and user object updates

### Description

Refactors our Statsig initialization and user object updating to ensure the following:
1. `startOnboardingTime` is available to statsig on app launch.
2. _Meaningful_ fetches to Statsig will never occur while the SDK is uninitialized.

See [this Slack thread](https://valora-app.slack.com/archives/C02KBT0DAHJ/p1680715814180279) and the following discussion for more information and context on the approach used here.

What this does/some more context:

Previously, we were updating the Statsig user object whenever a user trait changes, in the background. This can happen at any time, asynchronously. When updating the Statsig user, the Statsig SDK is briefly re-initialized, during which experiment fetching fails due to the SDK being uninitialized. This meant that there was a race condition where experiments could return default values if fetched while a user trait was being updated.

Additionally, when initializing the Statsig SDK, and while updating the Statsig user object, we were not passing _all_ properties, just the ones we wanted to update. `Statsig.updateUser` actually functions as a `PUT` operation instead of a `PATCH`, so we need to specify all properties we want, _every time_ we call `Statsig.updateUser`.

To solve the first problem, we now _explcitly_ call `Statsig.updateUser` at key moments and _block on its completion_ instead of asynchronously. Namely, when:
1. `startOnboardingTime` is determined for the first (and only) time.
2. Onboarding is completed, and the user's `walletAddress` is guaranteed to exist.

`Statsig.updateUser` is _never called_ after onboarding is complete. It is called _exactly_ twice for a user, ever. For 1., we block on `Statsig.updateUser` _before_ navigating to the first onboarding page that requires experiment parameters, in order to ensure that the correct parameters are fetched. For 2., when onboarding is completed, we fire off a new action that triggers a saga responsible for calling `Statsig.updateUser`, waiting for it to finish, and navigating to a specified post-onboarding screen (which is assumed to depend on the Statsig user object being up-to-date). Because the navigate necessarily occurs after the `Statsig.updateUser` resolves, the post-onboarding screen will always fetch the correct parameters, and not result in uninitialized fetches.

This approach of using a saga _does present_ the possibility of a small number of non-meaningful fetches to Statsig being made in the uninitialized state. This is because the last screen in onbaording will _always_ depend on Statsig experiment params. WLOG, consider the Phone Verification screen. When a user presses the "Skip" button, we kick off the saga to update the Statig user object and navigate to the post-onboarding screen. Pressing the "Skip" button _also_ induces state change, which triggers a re-render. The Phone Verification screen relies `getOnboardingStepValues` method in order to determine the steps to display in the header, which in turn relies on a fetch to Statsig to get experiment params. If this re-render occurs _while_ `Statsig.updateUser` is being called via the saga, the SDK will be briefly uninitialized, and the calls made to Statsig while re-determining onboarding steps for the re-render will return default values due to the SDK being uninitialized. The end-user impact of this is that there is a _chance_ that step numbers may (very) briefly be incorrect between the "Skip" button press and the navigate to the post-onboarding screen. This is something that's tolerable for the time being, though not ideal, and can likely be avoided when we make the switch to using the RN-specific `StatsigProvider` pattern.

To solve the _second_ problem from above (not passing all properties to Statsig `initialize` and `updateUser`), we've done two things:
1. When initializing the Statsig SDK, fetch all relevant parameters needed for the user object from Redux.
2. Wrap all calls to `Statsig.updateUser` in a `PATCH`-like interface, which merges default values from Redux with optional overrides from the caller.

### Test plan

Unit and manual tested. I would appreciate if reviewers could test this locally by performing the following steps:

1. Pull this branch.
2. Build/run a dev build.
3. In the dev build, reset Valora.
4. Open Flipper (or whatever you use to monitor logs) and filter for "Statsig".
5. Begin onboarding, and stop on the Phone Verification screen. It does not matter what experiment variants you are bucketed into. You should not see any logs.
6. Press the "Skip" button. You should see an action dispatched, and potentially, a small number of logs warning about some uninitialized calls to Statsig (as mentioned above).
7. Mess around in the app briefly, close and re-open, etc. You should not see any more logs (ever), and should have the same experimental variants as before.

While testing, I would encourage you to follow along with the exposure events in the diagnostics view of the [recovery phrase experiment](https://console.statsig.com/4plizaPmWwPL21ASV4QAO0/experiments/recovery_phrase_in_onboarding/diagnostics) or the [choose your adventure experiment](https://console.statsig.com/4plizaPmWwPL21ASV4QAO0/experiments/choose_your_adventure/diagnostics) (make sure to enable non-production logs), and make sure that everything looks right to you.

In between steps 5. and 6., you can try closing the app and re-opening it; the behavior should be the same as mentioned above (we experienced some issues this during testing, but have fixed them; testing this is mostly just a sanity check).

### Related issues

N/A

### Backwards compatibility

Yes.
